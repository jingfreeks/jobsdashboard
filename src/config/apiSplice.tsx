/* eslint-disable @typescript-eslint/no-explicit-any */
import {setCredentials, setLogout} from '@/features/auth';
import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
const baseUrl = import.meta.env.VITE_API_BASE_URL;
const baseQuery = fetchBaseQuery({
  baseUrl,
  credentials: 'include',
   
  prepareHeaders: (headers, {getState}: any) => {
    const token = getState().auth.token;
    if (token) {
      headers.set('authorization', `Bearer ${token}`);
    }
    return headers;
  },
});

const baseQueryWithAuth = async (
  args: string,
  api: any,
  extraOptions: string,
) => {
  let result: any = await baseQuery(args, api, extraOptions);
  console.log(result);
  if (result?.error?.originalStatus === 403) {
    const refereshResult = await baseQuery('/refresh', api, extraOptions);

    if (refereshResult?.data) {
      const currentAuth = api.getState().auth;
      //store the new token
      api.dispatch(setCredentials({
        user: currentAuth.user,
        accessToken: (refereshResult.data as any).accessToken,
        userId: currentAuth.userId,
        roles: currentAuth.roles
      }));
      //retry the original query with new access token
      result = await baseQuery(args, api, extraOptions);
    } else {
      api.dispatch(setLogout());
    }
  }
  return result;
};

export const apiSlice = createApi<any, any>({
  baseQuery: baseQueryWithAuth,
  reducerPath: 'api',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  endpoints: _ => ({}),
  tagTypes: ['City', 'Jobs', 'MyJobs', 'JobDetails', 'Profile'] as any,
});