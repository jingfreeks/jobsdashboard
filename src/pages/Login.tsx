import { useState } from "react";
import { useNavigate, Link } from "react-router-dom";
import { useForm } from "react-hook-form";
import logo from "../assets/react.svg";
import { Eye, EyeOff } from "lucide-react";
import { useLoginMutation } from "../features/auth/loginApiSlice";

type LoginFormData = {
  email: string;
  password: string;
};

const Login = () => {
  const navigate = useNavigate();
  const [showPassword, setShowPassword] = useState(false);
  const { register, handleSubmit, formState: { errors } } = useForm<LoginFormData>();
  const [login, { isLoading, error }] = useLoginMutation();

  const onSubmit = async (formData: LoginFormData) => {
    try {
      await login(formData).unwrap();
      navigate("/dashboard");
    } catch {
      // Error is handled by RTK Query's error state
    }
  };

  return (
    <div className="min-h-screen flex items-center justify-center ">
      <div className="w-full max-w-md bg-white rounded-2xl shadow-2xl px-8 py-10 flex flex-col items-center">
        <img src={logo} alt="Logo" className="w-20 h-20 mb-6" />
        <h1 className="text-2xl font-bold mb-6 text-center">Log in to Your Account</h1>
        <form onSubmit={handleSubmit(onSubmit)} className="w-full flex flex-col gap-4">
          <div>
            <label htmlFor="email" className="block text-sm font-medium mb-1">Email</label>
            <input
              id="email"
              type="text"
              {...register("email", { required: "Email is required" })}
              className="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            />
            {errors.email && <span className="text-red-500 text-xs">{errors.email.message as string}</span>}
          </div>
          <div>
            <label htmlFor="password" className="block text-sm font-medium mb-1">Password</label>
            <div className="relative">
              <input
                id="password"
                type={showPassword ? "text" : "password"}
                {...register("password", { required: "Password is required" })}
                className="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600 pr-10"
              />
              <button
                type="button"
                className="absolute right-2 top-1/2 -translate-y-1/2 text-gray-400 hover:text-gray-600"
                onClick={() => setShowPassword((v) => !v)}
                tabIndex={-1}
              >
                {showPassword ? (
                  <EyeOff className="h-4 w-4" />
                ) : (
                  <Eye className="h-4 w-4" />
                )}
              </button>
            </div>
            {errors.password && <span className="text-red-500 text-xs">{errors.password.message as string}</span>}
          </div>
          {error && <div className="text-red-500 text-sm mb-2 text-center">{(error as any)?.data?.error?.message || "Invalid credentials"}</div>}
          <button type="submit" className="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition" disabled={isLoading}>
            {isLoading ? "Logging in..." : "Log in"}
          </button>
        </form>
        <div className="w-full flex flex-col items-center mt-4 gap-1">
          <a href="#" className="text-xs text-gray-500 hover:underline">Forgot password?</a>
          <span className="text-sm text-gray-700">Don't have an account? <Link to="/register" className="text-[#0856d1] hover:underline font-medium">Sign up</Link></span>
        </div>
      </div>
    </div>
  );
};

export default Login;