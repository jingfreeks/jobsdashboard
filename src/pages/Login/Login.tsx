import { useState } from "react";
import { useNavigate, Link } from "react-router-dom";
import { useForm } from "react-hook-form";
import logo from "@/assets/react.svg";
import { Eye, EyeOff } from "lucide-react";
import {useDispatch} from 'react-redux';
import { useLoginMutation } from "@/features/loginApiSlice";
import type { FetchBaseQueryError } from '@reduxjs/toolkit/query';
import { setCredentials } from "@/features/auth";
import type { AppDispatch } from "@/config/store";

type LoginFormData = {
  username: string;
  password: string;
};

const Login = () => {
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();
  const [showPassword, setShowPassword] = useState(false);
  const { register, handleSubmit, formState: { errors } } = useForm<LoginFormData>();
  const [login, { isLoading, error }] = useLoginMutation();

  const onSubmit = async (formData: LoginFormData) => {
    try {
      console.log(formData);
     const result  =  await login({
        username: formData.username,
        password: formData.password,
      }).unwrap();
      if (result && typeof result === "object") {
        dispatch(setCredentials({ ...(result as object), user: formData.username }));
      }
    navigate("/dashboard");
    } catch {
      // Error is handled by RTK Query's error state
    }
  };

  // Helper type guard
  function isFetchBaseQueryError(
    error: unknown
  ): error is FetchBaseQueryError {
    return typeof error === 'object' && error != null && 'status' in error;
  }

  let loginErrorMessage: string | null = null;
  if (error) {
    if (isFetchBaseQueryError(error)) {
      const msg = (error.data as { error?: { message?: string } })?.error?.message;
      loginErrorMessage = msg ? String(msg) : "Invalid credentials";
    } else {
      loginErrorMessage = "Invalid credentials";
    }
  }

  return (
    <div className="min-h-screen flex items-center justify-center ">
      <div className="w-full max-w-md bg-white rounded-2xl shadow-2xl px-8 py-10 flex flex-col items-center">
        <img src={logo} alt="Logo" className="w-20 h-20 mb-6" />
        <h1 className="text-2xl font-bold mb-6 text-center">Log in to Your Account</h1>
        <form onSubmit={handleSubmit(onSubmit)} className="w-full flex flex-col gap-4">
          <div>
            <label htmlFor="email" className="block text-sm font-medium mb-1">Email</label>
            <input
              id="email"
              type="text"
              {...register("username", { required: "Username is required" })}
              className="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            />
            {errors.username && <span className="text-red-500 text-xs">{errors.username.message as string}</span>}
          </div>
          <div>
            <label htmlFor="password" className="block text-sm font-medium mb-1">Password</label>
            <div className="relative">
              <input
                id="password"
                type={showPassword ? "text" : "password"}
                {...register("password", { required: "Password is required" })}
                className="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600 pr-10"
              />
              <button
                type="button"
                className="absolute right-2 top-1/2 -translate-y-1/2 text-gray-400 hover:text-gray-600"
                onClick={() => setShowPassword((v) => !v)}
                tabIndex={-1}
              >
                {showPassword ? (
                  <EyeOff className="h-4 w-4" />
                ) : (
                  <Eye className="h-4 w-4" />
                )}
              </button>
            </div>
            {errors.password && <span className="text-red-500 text-xs">{errors.password.message as string}</span>}
          </div>
          {loginErrorMessage && (
            <div className="text-red-500 text-sm mb-2 text-center">
              {loginErrorMessage}
            </div>
          )}
          <button type="submit" className="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition" disabled={isLoading}>
            {isLoading ? "Logging in..." : "Log in"}
          </button>
        </form>
        <div className="w-full flex flex-col items-center mt-4 gap-1">
          <a href="#" className="text-xs text-gray-500 hover:underline">Forgot password?</a>
          <span className="text-sm text-gray-700">Don't have an account? <Link to="/register" className="text-[#0856d1] hover:underline font-medium">Sign up</Link></span>
        </div>
      </div>
    </div>
  );
};

export default Login;