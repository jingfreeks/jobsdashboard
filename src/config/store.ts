import { configureStore } from '@reduxjs/toolkit';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import authReducer from '@/features/auth';
import {apiSlice} from './apiSplice';

// Configure persistence for auth reducer
const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['user', 'token', 'userId', 'roles'], // Only persist these fields
  timeout: 10000, // 10 second timeout for storage operations
  serialize: true, // Enable serialization
  deserialize: true, // Enable deserialization
};

const persistedAuthReducer = persistReducer(authPersistConfig, authReducer);

export const store = configureStore({
  reducer: {
    auth: persistedAuthReducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        ignoredPaths: ['auth'], // Ignore auth state for serialization checks
      },
      immutableCheck: {
        ignoredPaths: ['auth'], // Ignore auth state for immutability checks
      },
    }).concat(apiSlice.middleware),
  devTools: process.env.NODE_ENV !== 'production',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch; 