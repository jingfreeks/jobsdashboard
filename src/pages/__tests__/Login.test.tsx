import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import Login from '../Login/Login';
import { store } from '@/config/store';

// Mock useLoginMutation to avoid real API calls
vi.mock('@/features/loginApiSlice', () => ({
  useLoginMutation: () => [vi.fn(), { isLoading: false, error: undefined }],
}));

// Mock setCredentials and selectIsAuthenticated to avoid Redux side effects
vi.mock('@/features/auth', () => ({
  __esModule: true,
  default: () => ({}), // mock reducer
  setCredentials: vi.fn(),
  selectIsAuthenticated: vi.fn(() => false), // Mock as not authenticated by default
}));

describe('Login', () => {
  it('renders login form fields and button', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByLabelText(/Email/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Password/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Log in/i })).toBeInTheDocument();
    
    // Snapshot test
    expect(container).toMatchSnapshot();
  });

  it('shows validation error if fields are empty', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );
    fireEvent.click(screen.getByRole('button', { name: /Log in/i }));
    expect(await screen.findByText(/Username is required/i)).toBeInTheDocument();
    expect(await screen.findByText(/Password is required/i)).toBeInTheDocument();
    
    // Snapshot test after validation errors
    expect(container).toMatchSnapshot();
  });

  it('matches login form snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
}); 