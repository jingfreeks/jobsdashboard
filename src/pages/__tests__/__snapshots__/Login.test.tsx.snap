// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > matches login form snapshot 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center "
  >
    <div
      class="w-full max-w-md bg-white rounded-2xl shadow-2xl px-8 py-10 flex flex-col items-center"
    >
      <img
        alt="Logo"
        class="w-20 h-20 mb-6"
        src="/src/assets/react.svg"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Log in to Your Account
      </h1>
      <form
        class="w-full flex flex-col gap-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="email"
            name="username"
            type="text"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="password"
          >
            Password
          </label>
          <div
            class="relative"
          >
            <input
              class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600 pr-10"
              id="password"
              name="password"
              type="password"
            />
            <button
              class="absolute right-2 top-1/2 -translate-y-1/2 text-gray-400 hover:text-gray-600"
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-eye h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="3"
                />
              </svg>
            </button>
          </div>
        </div>
        <button
          class="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition"
          type="submit"
        >
          Log in
        </button>
      </form>
      <div
        class="w-full flex flex-col items-center mt-4 gap-1"
      >
        <a
          class="text-xs text-gray-500 hover:underline"
          href="#"
        >
          Forgot password?
        </a>
        <span
          class="text-sm text-gray-700"
        >
          Don't have an account? 
          <a
            class="text-[#0856d1] hover:underline font-medium"
            data-discover="true"
            href="/register"
          >
            Sign up
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Login > renders login form fields and button 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center "
  >
    <div
      class="w-full max-w-md bg-white rounded-2xl shadow-2xl px-8 py-10 flex flex-col items-center"
    >
      <img
        alt="Logo"
        class="w-20 h-20 mb-6"
        src="/src/assets/react.svg"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Log in to Your Account
      </h1>
      <form
        class="w-full flex flex-col gap-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="email"
            name="username"
            type="text"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="password"
          >
            Password
          </label>
          <div
            class="relative"
          >
            <input
              class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600 pr-10"
              id="password"
              name="password"
              type="password"
            />
            <button
              class="absolute right-2 top-1/2 -translate-y-1/2 text-gray-400 hover:text-gray-600"
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-eye h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="3"
                />
              </svg>
            </button>
          </div>
        </div>
        <button
          class="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition"
          type="submit"
        >
          Log in
        </button>
      </form>
      <div
        class="w-full flex flex-col items-center mt-4 gap-1"
      >
        <a
          class="text-xs text-gray-500 hover:underline"
          href="#"
        >
          Forgot password?
        </a>
        <span
          class="text-sm text-gray-700"
        >
          Don't have an account? 
          <a
            class="text-[#0856d1] hover:underline font-medium"
            data-discover="true"
            href="/register"
          >
            Sign up
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Login > shows validation error if fields are empty 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center "
  >
    <div
      class="w-full max-w-md bg-white rounded-2xl shadow-2xl px-8 py-10 flex flex-col items-center"
    >
      <img
        alt="Logo"
        class="w-20 h-20 mb-6"
        src="/src/assets/react.svg"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Log in to Your Account
      </h1>
      <form
        class="w-full flex flex-col gap-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="email"
            name="username"
            type="text"
          />
          <span
            class="text-red-500 text-xs"
          >
            Username is required
          </span>
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="password"
          >
            Password
          </label>
          <div
            class="relative"
          >
            <input
              class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600 pr-10"
              id="password"
              name="password"
              type="password"
            />
            <button
              class="absolute right-2 top-1/2 -translate-y-1/2 text-gray-400 hover:text-gray-600"
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-eye h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="3"
                />
              </svg>
            </button>
          </div>
          <span
            class="text-red-500 text-xs"
          >
            Password is required
          </span>
        </div>
        <button
          class="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition"
          type="submit"
        >
          Log in
        </button>
      </form>
      <div
        class="w-full flex flex-col items-center mt-4 gap-1"
      >
        <a
          class="text-xs text-gray-500 hover:underline"
          href="#"
        >
          Forgot password?
        </a>
        <span
          class="text-sm text-gray-700"
        >
          Don't have an account? 
          <a
            class="text-[#0856d1] hover:underline font-medium"
            data-discover="true"
            href="/register"
          >
            Sign up
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;
